lib/src/main/java/com/diffplug/spotless/generic/IdeaStep.java

```diff
@@ -294,14 +356,116 @@ private static class IdeaStepFormatterCleanupResources implements AutoCloseable
 		private final File uniqueBuildFolder;
 		@Nonnull
 		private final ProcessRunner runner;
+		private final int batchSize;
+
+		// Batch processing state (transient - not serialized)
+		private final Map<File, String> formattedCache = new LinkedHashMap<>();
+		private final List<State.FileToFormat> pendingBatch = new ArrayList<>();
+		private final Lock batchLock = new ReentrantLock();
+		private State currentState;
 
-		public IdeaStepFormatterCleanupResources(@Nonnull File uniqueBuildFolder, @Nonnull ProcessRunner runner) {
+		public IdeaStepFormatterCleanupResources(@Nonnull File uniqueBuildFolder, @Nonnull ProcessRunner runner, int batchSize) {
 			this.uniqueBuildFolder = uniqueBuildFolder;
 			this.runner = runner;
+			this.batchSize = batchSize;
+		}
+
+		/**
+		 * Formats a single file, using batch processing for efficiency.
+		 * Files are accumulated and formatted in batches to minimize IDEA process startups.
+		 */
+		String formatFile(State state, String unix, File file) throws Exception {
+			batchLock.lock();
+			try {
+				// Store the state reference for batch processing
+				if (currentState == null) {
+					currentState = state;
+				}
+
+				// Check if we already have the formatted result cached
+				if (formattedCache.containsKey(file)) {
+					String result = formattedCache.remove(file);
+					LOGGER.debug("Returning cached formatted result for file: {}", file);
+					return result;
+				}
+
+				// Create a temporary file for this content
+				File tempFile = Files.createTempFile(uniqueBuildFolder.toPath(), "spotless", file.getName()).toFile();
+				Files.write(tempFile.toPath(), unix.getBytes(StandardCharsets.UTF_8));
+
+				// Add to pending batch
+				pendingBatch.add(new State.FileToFormat(tempFile, file));
+				LOGGER.debug("Added file {} to pending batch (size: {})", file, pendingBatch.size());
+
+				// If batch is full, process it
+				if (pendingBatch.size() >= batchSize) {
+					LOGGER.info("Batch size reached ({}/{}), processing batch", pendingBatch.size(), batchSize);
+					processPendingBatch();
+				}
+
+				// Check cache again after potential batch processing
+				if (formattedCache.containsKey(file)) {
+					return formattedCache.remove(file);
+				}
+
+				// If still not in cache, we need to process immediately (shouldn't happen normally)
+				// This is a safety fallback
+				LOGGER.warn("File {} not found in cache after batch processing, forcing immediate format", file);
+				List<State.FileToFormat> singleFileBatch = new ArrayList<>();
+				singleFileBatch.add(new State.FileToFormat(tempFile, file));
+				currentState.formatBatch(this, singleFileBatch);
+				return formattedCache.remove(file);
+
+			} finally {
+				batchLock.unlock();
+			}
+		}
+
+		/**
+		 * Caches a formatted result for a file.
+		 */
+		void cacheFormattedResult(File originalFile, String formatted) {
+			formattedCache.put(originalFile, formatted);
+		}
```

Comment:
unfortunately, I don't think using the `File` alone is sufficient for a key. The key needs to include the file's content. This is because there are other steps which will operate on the file before or after the idea step.